src/particles.zig

```diff
@@ -0,0 +1,413 @@
+		textures.deinit();
+		emissionTextures.deinit();
+		textureArray.deinit();
+		emissionTextureArray.deinit();
+		particleTypeHashmap.deinit(arenaAllocator.allocator);
+		ParticleSystem.deinit();
+		arena.deinit();
+	}
+
+	pub fn register(assetsFolder: []const u8, id: []const u8, zon: ZonElement) void {
+		const animationFrames = zon.get(u32, "animationFrames", 1);
+		const textureId = zon.get([]const u8, "texture", "cubyz:spark");
+
+		const particleType = readTextureDataAndParticleType(assetsFolder, textureId, animationFrames);
+
+		particleTypeHashmap.put(arenaAllocator.allocator, id, @intCast(types.items.len)) catch unreachable;
+		types.append(particleType);
+
+		std.log.debug("Registered particle type: {s}", .{id});
+	}
+	fn readTextureDataAndParticleType(assetsFolder: []const u8, textureId: []const u8, animationFrames: u32) ParticleType {
+		var typ: ParticleType = undefined;
+
+		const base = readTexture(assetsFolder, textureId, ".png", Image.defaultImage, .isMandatory);
+		const emission = readTexture(assetsFolder, textureId, "_emission.png", Image.emptyImage, .isOptional);
+
+		typ.startFrame = @floatFromInt(textures.items.len);
+		typ.size = @as(f32, @floatFromInt(base.width))/16;
+		for(0..animationFrames) |i| {
+			textures.append(extractAnimationSlice(base, i, animationFrames, textureId, .base));
+
+			const emmisionSlice = if(emission.imageData.ptr != Image.emptyImage.imageData.ptr)
+				extractAnimationSlice(emission, i, animationFrames, textureId, .emmision)
+			else
+				Image.emptyImage;
+			emissionTextures.append(emmisionSlice);
+		}
+
+		typ.animationFrames = @floatFromInt(animationFrames);
+		return typ;
+	}
+
+	fn readTexture(assetsFolder: []const u8, textureId: []const u8, suffix: []const u8, default: graphics.Image, status: enum {isOptional, isMandatory}) graphics.Image {
+		var splitter = std.mem.splitScalar(u8, textureId, ':');
+		const mod = splitter.first();
+		const id = splitter.rest();
+
+		const gameAssetsPath = std.fmt.allocPrint(main.stackAllocator.allocator, "assets/{s}/particles/textures/{s}{s}", .{mod, id, suffix}) catch unreachable;
+		defer main.stackAllocator.free(gameAssetsPath);
+
+		const worldAssetsPath = std.fmt.allocPrint(main.stackAllocator.allocator, "{s}/{s}/particles/textures/{s}{s}", .{assetsFolder, mod, id, suffix}) catch unreachable;
+		defer main.stackAllocator.free(worldAssetsPath);
+
+		return graphics.Image.readFromFile(arenaAllocator, worldAssetsPath) catch graphics.Image.readFromFile(arenaAllocator, gameAssetsPath) catch {
+			if(status == .isMandatory) std.log.err("Particle texture not found in {s} and {s}.", .{worldAssetsPath, gameAssetsPath});
+			return default;
+		};
+	}
+
+	fn extractAnimationSlice(image: Image, frame: usize, frames: usize, imageName: []const u8, textureType: enum {base, emmision}) Image {
+		if(image.height%frames != 0) {
+			std.log.err("Particle texture size is not divisible by its frame count for {s} in {s} texture", .{imageName, @tagName(textureType)});
```

Comment:
I would suggest suggest including those numbers in the error message
```suggestion
			std.log.err("Particle texture size ({}) is not divisible by its frame count ({}) for {s} in {s} texture", .{image.height, frames, imageName, @tagName(textureType)});
```

Reply:
Yeah, textures should always be square.

Reply:
Hmm, what about rain?

Reply:
i guess the sides are going to be 0 alpha, tho i could try making particles have separate width and height sizes if needed

Reply:
but extra discarded pixels are probably not good

Reply:
> Hmm, what about rain?

What about it?

> but extra discarded pixels are probably not good

You will be discarding pixels for a lot of particles anyway, rarely any particle fills the whole square.

Reply:
> You will be discarding pixels for a lot of particles anyway, rarely any particle fills the whole square.

There is a big difference between a 25% discard rate for a circular texture and a >90% discard rate for rain.